src/psyclone/psyir/backend/sympy_writer.py

```diff
@@ -115,86 +132,212 @@ def __init__(self, type_map=None):
             self._intrinsic.add(op_str)
             self._op_to_str[operator] = op_str
 
-    @staticmethod
-    def create_type_map(list_of_expressions):
+    # -------------------------------------------------------------------------
+    def __new__(cls, *expressions):
+        '''This function allows the SymPy writer to be used in two
+        different ways: if only the SymPy expression of the PSyIR expressions
+        are required, it can be called as::
+
+            sympy_expressions = SymPyWriter(exp1, exp2, ...)
+
+        But if additional information is needed (e.g. the SymPy type map, or
+        to convert a SymPy expression back to PSyIR), an instance of the
+        SymPyWriter must be kept, e.g.::
+
+            writer = SymPyWriter()
+            sympy_expressions = writer([exp1, exp2, ...])
+            writer.type_map
+
+        :param expressions: a (potentially empty) tuple of PSyIR nodes
+            to be converted to SymPy expressions.
+        :type expressions: Tuple[:py:class:`psyclone.psyir.nodes.Node`]
+
+        :returns: either an instance of SymPyWriter, if no parameter is
+            specified, or a list of SymPy expressions.
+        :rtype: Union[:py:class:`psyclone.psyir.backend.SymPyWriter`,
+                      List[:py:class:`sympy.core.basic.Basic`]
+
         '''
-        This function creates a dictionary mapping each Reference in any
-        of the expressions to either a Sympy Function (if the reference
+        if expressions:
+            # If we have parameters, create an instance of the writer
+            # and use it to convert the expressions:
+            writer = SymPyWriter()
+            return writer(expressions)
+
+        # No parameter, just create an instance and return it:
+        return super().__new__(cls)
+
+    # -------------------------------------------------------------------------
+    def __getitem__(self, _):
+        '''This function is only here to trick pylint into thinking that
```

Comment:
That being so, should it raise `NotImplementedError` or `InternalError` as we'd never expect it to be called?

Reply:
Thank you. People debugging in the future may thank you too :-)